lightning/src/ln/channel.rs

```diff
@@ -1684,6 +1685,103 @@ pub(super) trait InteractivelyFunded<SP: Deref> where SP::Target: SignerProvider
+			}
+		}
+
+		let total_input_satoshis: u64 = funding_inputs.iter().map(
+			|input| input.1.as_transaction().output.get(input.0.previous_output.vout as usize).map(|out| out.value.to_sat()).unwrap_or(0)
+		).sum();
+		if total_input_satoshis < self.dual_funding_context().our_funding_satoshis {
+			return Err(APIError::APIMisuseError {
+				err: format!("Total value of funding inputs must be at least funding amount. It was {} sats",
+					total_input_satoshis) });
+		}
+
+		// Add output for funding tx
+		let mut funding_outputs = Vec::new();
+		let funding_output_value_satoshis = self.context().get_value_satoshis();
+		let funding_output_script_pubkey = self.context().get_funding_redeemscript().to_p2wsh();
+		let expected_remote_shared_funding_output = if self.is_initiator() {
+			let tx_out = TxOut {
+				value: Amount::from_sat(funding_output_value_satoshis),
+				script_pubkey: funding_output_script_pubkey,
+			};
+			funding_outputs.push(
+				if self.dual_funding_context().their_funding_satoshis.unwrap_or(0) == 0 {
+					OutputOwned::SharedControlFullyOwned(tx_out)
+				} else {
+					OutputOwned::Shared(SharedOwnedOutput::new(
+						tx_out, self.dual_funding_context().our_funding_satoshis
+					))
+				}
+			);
+			None
+		} else {
+			Some((funding_output_script_pubkey, funding_output_value_satoshis))
+		};
+
+		// Optionally add change output
+		if let Some(change_value) = calculate_change_output_value(
+			self.is_initiator(), self.dual_funding_context().our_funding_satoshis,
+			&funding_inputs_prev_outputs, &funding_outputs,
+			self.dual_funding_context().funding_feerate_sat_per_1000_weight,
+			self.context().holder_dust_limit_satoshis,
+		) {
+			let change_script = signer_provider.get_destination_script(self.context().channel_keys_id).map_err(
+				|err| APIError::APIMisuseError {
+					err: format!("Failed to get change script as new destination script, {:?}", err),
+				})?;
+			add_funding_change_output(change_value, change_script,
+				&mut funding_outputs, self.dual_funding_context().funding_feerate_sat_per_1000_weight);
+		}
+
+		let constructor_args = InteractiveTxConstructorArgs {
+			entropy_source,
+			holder_node_id,
+			counterparty_node_id: self.context().counterparty_node_id,
+			channel_id: self.context().channel_id(),
+			feerate_sat_per_kw: self.dual_funding_context_mut().funding_feerate_sat_per_1000_weight,
+			is_initiator: self.is_initiator(),
+			funding_tx_locktime: self.dual_funding_context_mut().funding_tx_locktime,
+			inputs_to_contribute: funding_inputs,
+			outputs_to_contribute: funding_outputs,
+			expected_remote_shared_funding_output,
+		};
+		let mut tx_constructor = InteractiveTxConstructor::new(constructor_args)
+			.map_err(|_| APIError::APIMisuseError { err: "Incorrect shared output provided".into() })?;
+		let msg = tx_constructor.take_initiator_first_message();
+
+		self.interactive_tx_constructor_mut().replace(tx_constructor);
```

Comment:
Could just assign instead of using `replace`. Do we ever care about the overridden value? Will it always be `None`?

Reply:
I changed to assign (with a dereference of the mut ref on the left side), I'm not sure why originally a `replace` was used (@dunxen ?)

Reply:
Not sure, was probably a half-baked idea as old value is not used anyway. Assign is good.